src/crimson/osd/heartbeat.cc

```diff
@@ -50,9 +50,9 @@ seastar::future<> Heartbeat::start(entity_addrvec_t front_addrs,
 
   using crimson::net::SocketPolicy;
   front_msgr->set_policy(entity_name_t::TYPE_OSD,
-                         SocketPolicy::stateless_server(0));
+                         SocketPolicy::lossless_peer(0));
   back_msgr->set_policy(entity_name_t::TYPE_OSD,
-                        SocketPolicy::stateless_server(0));
+                        SocketPolicy::lossless_peer(0));
```

Comment:
I don't think so, this will break the compatibility to be able to talk with classic OSDs.

> this would cause problem
when both sides of a heartbeat connection tries to reconnect, which is not expected for
lossy connections

Could you elaborate what you have observed if it is an issue, does it cause serious bugs?

Reply:
stateless_server's connection is lossy, so when both sides tries to reconnect, there's a high possibility that the code goes into handle_existing_connection(), and as the connection is lossy, one of them would be reset and the surviving connection goes into execute_establishing() method in which it tries to register the connection. But if when handling reset, the application tries to reconnect again, then the connection register would trigger the assert "ceph_assert(added);" in SocketMessenger::register_conn.

I agree with on that this breaks compatibility with classic osds, shall I add two hb client msgr just like what the classic osds do? Thanks:-)

Reply:
> when handling reset, the application tries to reconnect again, then the connection registering would trigger the assert "ceph_assert(added);" in SocketMessenger::register_conn.

Ahh, it's actually a bug which will also happen elsewhere. The reset event dispatching here should be asynchronous, this connection should be closed/unregistered and replaced before notifying reset and reconnect. Then the reconnect attempt will simply get the replaced connection.

> I agree with you on that this breaks compatibility with classic osds, shall I add two hb client msgrs just like what the classic osds do? Thanks:-)

Let me fix the bug first and see if there are other issues. According to my current understanding, it could be possible to save a connection between every two peering crimson OSDs.

Reply:
> EDIT: I think the policy should be SocketPolicy::lossy_client(0), as I see the differences between SocketPolicy::stateless_server(0) are:

Can lossy_client serve as a server?

Reply:
> Can lossy_client serve as a server to be connected to?

I think so. The only difference between `SocketPolicy::lossy_client(0)` and `SocketPolicy::stateless_server(0)` is the boolean `Policy::server`, which leads to some limited different behaviors as I listed above according to `src/crimson/net/ProtocolV2.cc`. And it seems to me that `SocketPolicy::stateless_server(0)` is not a correct choice.

Anyway, we can always fall back to the two-connection implementation, if `SocketPolicy::lossy_client(0)` turns out to be impractical.

Reply:
Could you split the commit "crimson/heartbeat: avoid heartbeat deadlock on connection reset" into two parts? The policy change is not related to the title.

And if you feel uncertain about the policy change, you can choose not to include it.

Reply:
@cyx1231st Ok, I'll split the commit:-)

Reply:
Did some tests with `SocketPolicy::lossy_client(0)`, no issue found. See https://github.com/ceph/ceph/pull/33909/commits/096d6d407367211f874f34d336b7c930322a3ca3